### ISSUE TITLE
ColumnWidth affectes table width

### ISSUE THREAD

**Issue by prabusendhil** [simonbengtsson/jsPDF-AutoTable#208]
Proof for ColumnWidth issue:
[document.pdf](https://github.com/simonbengtsson/jsPDF-AutoTable/files/672403/document.pdf

Issue 1: I am using columnWith for Employee Name.it takes table with out of box

Issue 2: With open saved pdf header border is not showing but view in new tab in browser it shows border right is availabe.
Code here:
```js
margin: {top: 70},
styles: {
    cellPadding: 3,
        fontSize: 8,
        valign: 'middle',

        overflow: 'linebreak',
        tableWidth: 'auto',
        //fileColor: [30, 30, 30],
        lineWidth: 1,
},
headerStyles: {
    //columnWidth: 'wrap',
    cellPadding: 2,
        lineWidth: 1,
        valign:'top',
        fontStyle: 'bold',
        halign: 'left',    //'center' or 'right'
        //fillColor: [250, 152, 95],
        //textColor: [78, 53, 73], //Black     
        textColor: [255, 255, 255], //White     
        fontSize: 8,

        rowHeight:20
},
```

**Comment by simonbengtsson** [comment]
When I saw your example I remembered that I have thought about this before. The reason this is happening is that the plugin does not know which column to shrink since all of them are shrunk maximally without hiding their content. What would you say should be the expected behavior here? Hide parts of the titles?

I'm not sure I understand issue 2. Can you elaborate?

**Comment by prabusendhil** [comment]
For issue 1 is there any fault in my code ?

Beacause i got directly from html. If i am not giving columnWidth is shows perfectly.approximately marigin right 40 space.If i am using columnWidth to particular column this is happend. Is there any trick?

For issue 2 if i see the output in browser shows the header have border. But some cells in header doesn't have border. While printing the border is not showing in header part alown.

**Comment by simonbengtsson** [comment]
1. Can you post the html table that causes the 40px space on each side issue?
2. Border and other styles will not be parsed from html tables. Use the `lineWidth` and `lineColor` styles ie `doc.autoTable(columns, rows, {styles: lineWidth: 1})`
3. What are you trying to accomplish? The autoTableAddPageContent was added in v2.1.

**Comment by prabusendhil** [comment]
Thank u Simons.
I want  to make row per page 20 only. But now it shows 35 rows per page. So that i am using
drawRow:function(row){
if((row.index >0 && row.index % 20) === 0){
doc.autoTableAddPageContent(pageContent);
} 
output: autoTableAddPageContent( undefined 

previously I used, 
if((row.index >0 && row.index % 20) === 0){
doc.autoTableAddPage();
}

output:  the page header and footer is not display in first page .from second page only it show
I have  v2.1. in my project.. .
is there any way to do this ?

**Comment by prabusendhil** [comment]
Thank a lot Simon.
Issue 1 and 2 is fixed while i am using lineColor and lineWidth
need help for below issue 
I want  to make row per page 20 only. But now it shows 35 rows per page. So that i am using
drawRow:function(row){
if((row.index >0 && row.index % 20) === 0){
doc.autoTableAddPageContent(pageContent);
[Attendance_Muster_Report.pdf](https://github.com/simonbengtsson/jsPDF-AutoTable/files/672542

} 
output: autoTableAddPageContent( undefined 

previously I used, 
if((row.index >0 && row.index % 20) === 0){
doc.autoTableAddPage();
}

output:  the page header and footer is not display in first page .from second page only it show
I have  v2.1. in my project.. .
is there any way to do this ?

**Comment by prabusendhil** [comment]
Issue 1 :  I uploaded two document one is without using columnWidth other one is using columnWidth for Employee name. If i used columnWidth table right portion extent to edge .plz compare with first uploaded document.still yoi can't get my point?

Issue 2: Is there any quick fix for per page count. Or else i when 3.0 will come to our hand Simon

**Comment by simonbengtsson** [comment]
Issue 1: The first document you uploaded without `columnWidth` looks good to me. Is it something wrong with it?

Issue 2 and 3: If you have a problem with images and jspdf I suggest posting a question on stackoverflow with the `jspdf` tag.

**Comment by prabusendhil** [comment]
Issue 1: exactly first document looms good to me also. But i need to give coulmnWidth for Employye Name. If i used columnWidth the table structure changed like ehat is in my second uploaded document Simon. There tge issue is if compare first and second document rightside space is missing .but i need equal space of left and right sides

**Comment by simonbengtsson** [comment]
Oh! Fair enough. You need to manually limit the width of the other columns since there is no room for the plugin to automatically shrink them. 

I have been thinking of what do here, but let the user handle it in the end. There is no obvious behavior to me, simply hide parts of the other columns?

**Comment by prabusendhil** [comment]
Thank a lot for your support Simon.
Issue 1: I think you got my exact problem. But I dont know how many columns will come to display. Because , these all are dynamic columns. I need to gIve column width for employee name only.so that i gave column width for 1st position element. Other column will display with auto column width. So that i build code lik this. It is required scenario. Please, do you think is there any trick to do this.it is very helpfull to me Simon.

Issue 2: I will post issue in stock overflow with jspdg denotion.

**Comment by simonbengtsson** [comment]
What do you want to happen to the other columns? Hide there content in favour of the employee name column?

**Comment by prabusendhil** [comment]
Other columns should take column auto property. The table alignment should not change lije 2nd uploaded image.table right side i need space like leftside i think leftside default margin value is 40 same i need right side

**Comment by simonbengtsson** [comment]
They are set to auto. But the plugin can't shrink those columns further without hiding the title of the columns. If you change the column titles you will see that they will shrink.

**Comment by prabusendhil** [comment]
Not an issue about shrink the columns .my issue only right side space is shrink. I need space in right side like left side

**Comment by prabusendhil** [comment]
No issue in columns simon. My issue is outside of table

**Comment by simonbengtsson** [comment]
Yes, but to make that space the plugin would need to shrink columns right? Maybe I am still misunderstanding something.

**Comment by prabusendhil** [comment]
[Attendance_Muster_Report-1.pdf](https://github.com/simonbengtsson/jsPDF-AutoTable/files/674954/Attendance_Muster_Report-1.pdf)
 
In this uploaded document i draw red mark  in right side space. Uplaoded document  have columnWidth of auto (all field s)

If i am using static columnWidth for employee name table red mark space occupied by last column. What is my expected output here. Except employee name all columns may shrink.but that columns should not comes to that red mark portion Simon.
Can u get my problem Simon?

**Comment by simonbengtsson** [comment]
Haha that was how I first interpreted your question. However as I have said I am not sure what the plugin should do to accomplish it as the other columns can't be smaller without hiding parts of their title.

**Comment by prabusendhil** [comment]
Thank you Simons

**Comment by simonbengtsson** [comment]
For the issue above regarding 20 rows per page. If you use `autoTableAddPage` instead of `autoTableAddPageContent` the code you posted will work:

```js
drawRow: function(row) {
    if (row.index > 0 && row.index % 20 === 0) {
        doc.autoTableAddPage(); // And not autoTableAddPageContent
    }
}
```

**Comment by prabusendhil** [comment]
If I use  code like above,  the header and footer render from second page ...

**Comment by simonbengtsson** [comment]
How are you adding the header and footer?

**Comment by prabusendhil** [comment]
var pageContent = function (data) {

                    // HEADER
                    doc.setFontSize(18);
                    doc.setTextColor(40);
                    doc.setFontStyle('normal');
                    
                    doc.addImage(options.data.datas.loadedImg, 'PNG', data.settings.margin.left, 10, 75, 25);
                    if (options.data.location) {
                        var Branch = "Branch - " + options.data.location;
                        doc.setFontSize(9);
                        doc.setTextColor(40);
                        doc.text(Branch, data.settings.margin.left, 50);
                    }

                    if (options.data.startDate) {
                        var startDate = "Start Date - " + options.data.startDate;
                        doc.setFontSize(9);
                        doc.setTextColor(40);
                        var xOffset = (doc.internal.pageSize.width / 2) - (doc.getStringUnitWidth(startDate) * doc.internal.getFontSize() / 2);
                        doc.text(startDate, xOffset, 50);
                    }

//                    if (options.data.financialYear) {
//                        var financialYear = "Financial Year - " + options.data.financialYear;
//                        doc.setFontSize(9);
//                        doc.setTextColor(40);
//                        var xOffset = (doc.internal.pageSize.width / 2) - (doc.getStringUnitWidth(financialYear) * doc.internal.getFontSize() / 2);
//                        doc.text(startDate, xOffset, 50);
//                    }
                    if (options.data.month) {
                        var month = "Month - " + options.data.month;
                        doc.setFontSize(9);
                        doc.setTextColor(40);
                        var xOffset = (doc.internal.pageSize.width / 2) - (doc.getStringUnitWidth(month) * doc.internal.getFontSize() / 2);
                        doc.text(month, xOffset, 50);
                    }
                    if (options.data.endDate) {
                        var endDate = "End Date - " + options.data.endDate;
                        doc.setFontSize(9);
                        doc.setTextColor(40);
                        var xOffset = (doc.internal.pageSize.width - 10) - (doc.getStringUnitWidth(endDate) * doc.internal.getFontSize() / 2);
                        doc.text(endDate, xOffset - 80, 50);
                    }
                    if (options.data.year) {
                        var year = "Year - " + options.data.year;
                        doc.setFontSize(9);
                        doc.setTextColor(40);
                        var xOffset = (doc.internal.pageSize.width - 10) - (doc.getStringUnitWidth(year) * doc.internal.getFontSize() / 2);
                        doc.text(year, xOffset - 60, 50);
                    }

                    if (options.data.financialYear) {
                        var year = "Financial Year - " + options.data.financialYear;
                        doc.setFontSize(9);
                        doc.setTextColor(40);
                        var xOffset = (doc.internal.pageSize.width - 10) - (doc.getStringUnitWidth(year) * doc.internal.getFontSize() / 2);
                        doc.text(year, xOffset - 90, 50);
                    }


                    var text = options.data.heading;
                    var xOffset = (doc.internal.pageSize.width / 2) - (doc.getStringUnitWidth(text) * doc.internal.getFontSize() / 2);
                    doc.setFontSize(13)
                    doc.text(text, xOffset - 10, 30);

                    // FOOTER
                    var str = "Page " + data.pageCount;
                    // Total page number plugin only available in jspdf v1.0+
                    if (typeof doc.putTotalPages === 'function') {
                        str = str + " of " + totalPagesExp;
                    }
                    doc.setFontSize(10);
                    if (options.data.datas.city)
                        doc.text("Copyright (c) Innoart Technologies. All rights reserved.", data.settings.margin.left, doc.internal.pageSize.height - 30);
                    var xOffset = (doc.internal.pageSize.width - 10) - (doc.getStringUnitWidth(str) * doc.internal.getFontSize() / 2);
                    if (!options.data.datas.city)
                        doc.text("Copyright (c) Innoart Technologies. All rights reserved.", data.settings.margin.left, doc.internal.pageSize.height - 30);
                    var xOffset = (doc.internal.pageSize.width - 10) - (doc.getStringUnitWidth(str) * doc.internal.getFontSize() / 2);
                    doc.text(str, xOffset, doc.internal.pageSize.height - 30);

                };

                doc.autoTable(res.columns, res.data, {
                    theme: 'striped',
                    //startY: doc.autoTableEndPosY() + 20,
                    addPageContent: pageContent,
//                    createdHeaderCell: function (cell, data) {
//                        console.log("cell")
//                        console.log(cell)
//                        console.log("data")
//                        console.log(data)
//                        if (cell.raw === 'Name') {
//                            cell.styles.fontSize = 15;
//                            cell.styles.textColor = 111;
//                        } else {//else rule for drawHeaderCell hook
//                            cell.styles.textColor = 255;
//                            cell.styles.fontSize = 10;
//
//                        }
//                    },

**Comment by prabusendhil** [comment]
The above manner I added header and footer Simon
